plugins/linguist-backend/README.md

```diff
@@ -112,6 +113,101 @@ return createRouter(
 
+## Linguist Tags Processor
+
+The `LinguistTagsProcessor` can be added into your catalog builder as a way to incorporate the language breakdown from linguist as `metadata.tags` on your entities. Doing so enables the ability to easily filter for entities in your catalog index based on the language of the source repository.
+
+### Processor Setup
+
+Setup the linguist tag processor in `packages/backend/src/plugins/catalog.ts`.
+
+```ts
+import { LinguistTagsProcessor } from '@backstage/plugin-linguist-backend';
+// ...
+export default async function createPlugin(
+  // ...
+  builder.addProcessor(
+    new LinguistTagsProcessor({
+      logger: env.logger,
+      discovery: env.discovery,
+    }),
+  );
+```
+
+### Processor Options
+
+#### `shouldProcessEntity`
+
+The `shouldProcessEntity` is a function you can pass into the processor which determines which entities should have language tags fetched from linguist and added to the entity. By default, this will only run on entities of `kind: Component`, however this function let's you fully customize which entities should be processed.
+
+As an example, you may choose to extend this to support both `Component` and `Resource` kinds along with allowing an opt-out annotation on the entity which entity authors can use.
+
+```ts
+new LinguistLanguageTagsProcessor({
+  logger: env.logger,
+  discovery: env.discovery,
+  shouldProcessEntity: (entity: Entity) => {
+    if (
+      ['Component', 'Resource'].includes(entity.kind) &&
+      !entity.metadata.annotations?.['backstage.io/ignore-linguist-tags']
+    ) {
+      return true;
+    }
+    return false;
+  },
+});
+```
+
+#### `tagTransformer`
+
+The `tagTransformer` option allows you to supply a custom function that can transform a linguist supplied language tag before it is appended to the entity tags. You can also use it to ignore certain language tags entirely by transforming the tag to `undefined`.
+
+The `defaultTagTransformer` is exported for your convenience to use in a chained transformer. It is recommended to still use it, as the default transformer will ensure that your language is valid per the [tags requirements](https://backstage.io/docs/features/software-catalog/descriptor-format/#tags-optional).
+
+As an example, one might use this supplied function to supply an allow list of languages.
+
+```ts
+new LinguistLanguageTagsProcessor({
+  logger: env.logger,
+  discovery: env.discovery,
+  tagTransformer: tag => {
+    const cleaned = defaultTagTransformer(lang);
+    return ['python', 'java', 'typescript'].includes(cleaned)
+      ? cleaned
+      : undefined;
+  },
+});
+```
+
+#### `cacheTTL`
+
+The `cacheTTL` option allows you to determine for how long this processor will cache languages for an `entityRef` before refreshing from the linguist backend. As this processor will run continuously, this cache is supplied to limit the load done on the linguist DB and API.
```

Comment:
I think some more research is needed before determining whether the cache is needed / not needed. I have not run this processor at large scale, and will admit that I created this cache to be a bit defensive in the situation of large catalogs that are constantly churning their entities. Each run of the processor loop would be fetching individual language breakdowns for every component entity, which is a lot of extra db reads for unchanging data.

In a 30 min default, I was trying to balance eventual consistency for relieving reads on DBs, but obviously that assumes that the reads are expensive enough in the first place to warrant that caching complexity and spend of extra process memory footprint (I assume, fairly small here as each entity probably only has a handful of languages associated, we may be talking MBs here).

>Values that are set during a processing run will only be visible in the directly
following run. The cache will be overwritten every run unless no new cache items
are written, in which case the existing values remain in the cache.

I'd like to know more about the above. What is the intention behind "values will only be visible in the directly following run"? Is this to make sure that caches are guaranteed to be bust to prevent growing memory? If I made sure to explicitly call `cache.set()` on each run with the previous cached values, would i be able to perpetuate the cache across several runs?

Reply:
re whether caching is needed or not. I mainly wanted to raise the question of whether we really need it. @awanlin maybe you can give a better opinion as the author of the linguist plugins.